### DISCUSSION TITLE
code understanding question in typecore.ml

### DISCUSSION THREAD

**Discussion by gasche** [ocaml/ocaml#12522] [Q&A]
I would like to understand the purpose of the following `(_ as x)` device in typecore.ml:

https://github.com/ocaml/ocaml/blob/b795460a99a00b6eb97c48f7bbe138f88dc47ef7/typing/typecore.ml#L1900-L1917

Question: what is the purpose of the `Tpat_var` special case?

cc @garrigue (and maybe @t6s, who worked on `type_pat` quite a bit recently?)

## Context

I am asking because this special case, combined with #12119, results in a change of typedtree emitted between 5.0 and the 5.1 release branch or trunk. Consider the following program

```ocaml
let x : int = 42
```

With 5.0, the typedtree generated corresponds to a `(x : int)` pattern (more precisely, pseudo-code would be `{ pat = Ppat_var "x"; extra = Ppat_constraint "int" }`).
With 5.1 or trunk, the typedtree generated corresponds to a `((_ as x) : int)` pattern (`{ pat = Ppat_alias (Ppat_any, Ppat_var "x"); extra = Ppat_constraint "int" }`).

This difference is causing me some minor issues down the line due to code that pattern-matches on the typedtree and has a special shortcut for `Ppat_var`, that ultimately interacts with code that @trefis and myself are writing to fix #7241.

## Observation

If I remove this special case, there is exactly one failure in the compiler testsuite. The following test:

```ocaml
class type numeral = object method fold : ('a -> 'a) -> 'a -> 'a end
class zero = object (_ : #numeral) method fold f x = x end
class next (n : #numeral) =
  object (_ : #numeral) method fold f x = n#fold f (f x) end
;;
[%%expect {|
class type numeral = object method fold : ('a -> 'a) -> 'a -> 'a end
class zero : object method fold : ('a -> 'a) -> 'a -> 'a end
class next : #numeral -> object method fold : ('a -> 'a) -> 'a -> 'a end
|}];;
```

now behaves as follows in `-principal` mode:

```
Line 4, characters 42-48:
|   object (_ : #numeral) method fold f x = n#fold f (f x) end
                                            ^^^^^^
Warning 18 [not-principal]: this use of a polymorphic method is not principal.
```

I am not sure what is causing the difference, and whether this is the intended effect of the `(_ as x)` wrapping, or it was meant to change something else somewhere else (that may or may not still exists in the compiler today; the wrapping was introduced in the development branch of 4.00).

If I understand correct, the warning comes from a difference in type-checking between the pattern `(n : #numeral)` and the pattern `((_ as n) : #numeral)` (which only occurs in the wrapping is disabled in the type-checker).

**Comment by lpw25** [comment]
I don't remember the exact details, but this is a hack to get the levels right by either removing some sharing or avoiding some traversal that lowers levels. I was able to remove it in my patch to add polymorphic parameters by fixing the treatment of levels in `type_let`. This also allowed me to remove the special case for handling constraints with polymorphic types in `type_pat`.

I think it requires [this commit](https://github.com/ocaml-flambda/ocaml-jst/pull/14/commits/cc6b75344edec3b9419168178e554992638f728c) to adjust `type_let` and then the parts of [this commit](https://github.com/ocaml-flambda/ocaml-jst/pull/14/commits/fd60674496200baf77889e1a25017d0a027885da) that are related to `type_pat`.
